modules/tensor_mechanics/include/materials/ComputeThermalExpansionEigenstrainBase.h

```diff
@@ -37,9 +38,28 @@ class ComputeThermalExpansionEigenstrainBase
    *                         expansion (derivative of thermal_strain wrt
    *                         temperature
    */
-  virtual void computeThermalStrain(Real & thermal_strain, Real & instantaneous_cte) = 0;
+  virtual void computeThermalStrain(Real & thermal_strain, Real & instantaneous_cte);
+  virtual void computeThermalStrain(ADReal & thermal_strain);
+
+  /// lag temperature variable
+  const bool _use_old_temperature;
+
+  /// old temperature for use_old_temperature = true
+  const VariableValue & _temperature_old;
+
+  /// current temperature
+  const GenericVariableValue<is_ad> & _temperature;
+
+  /// the eigenstrain derivative
+  typename std::conditional<is_ad, void *, MaterialProperty<RankTwoTensor> &>::type
```

Comment:
ah this was the step I could never figure out...

Reply:
A property pointer rather than a reference would be easier, but it would require changes in derived classes. Let me investigate the fallout...

Reply:
yeah, a few objects in Bison would require fixing. maybe worth it though.